### PR TITLE
llvm: try to build with gcc-5

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by danielnachun** [comment]
I saw in the errors a reference to the `-fpermissive` flag.  I tried setting that locally in Docker with `CMAKE_CXX_FLAGS` and `CMAKE_C_FLAGS` and it got past the first failure point.  Let's see now if can finish. 🤞

**Comment by danielnachun** [comment]
We tripped at the finish line ☹️.  @chandlerc Any thoughts on why this is happening?  I'm going to do a local test where I disable `lldb` to see if that is the only thing that is failing.
```
2021-06-15T08:13:15.3944958Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp: In constructor ‘lldb_private::process_linux::NativeRegisterContextLinux_x86_64::NativeRegisterContextLinux_x86_64(const lldb_private::ArchSpec&, lldb_private::NativeThreadProtocol&)’:
2021-06-15T08:13:15.3949081Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:296:34: error: use of deleted function ‘lldb_private::process_linux::NativeRegisterContextLinux::NativeRegisterContextLinux()’
2021-06-15T08:13:15.3950860Z        m_reg_info(), m_gpr_x86_64() {
2021-06-15T08:13:15.3951265Z                                   ^
2021-06-15T08:13:15.3952645Z In file included from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.h:14:0,
2021-06-15T08:13:15.3954646Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:11:
2021-06-15T08:13:15.3958157Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Linux/NativeRegisterContextLinux.h:18:7: note: ‘lldb_private::process_linux::NativeRegisterContextLinux::NativeRegisterContextLinux()’ is implicitly deleted because the default definition would be ill-formed:
2021-06-15T08:13:15.3960327Z  class NativeRegisterContextLinux
2021-06-15T08:13:15.3960889Z        ^
2021-06-15T08:13:15.3963572Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Linux/NativeRegisterContextLinux.h:18:7: error: no matching function for call to ‘lldb_private::NativeRegisterContextRegisterInfo::NativeRegisterContextRegisterInfo()’
2021-06-15T08:13:15.3966516Z In file included from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Linux/NativeRegisterContextLinux.h:12:0,
2021-06-15T08:13:15.3968539Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.h:14,
2021-06-15T08:13:15.3970531Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:11:
2021-06-15T08:13:15.3974350Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h:24:3: note: candidate: lldb_private::NativeRegisterContextRegisterInfo::NativeRegisterContextRegisterInfo(lldb_private::NativeThreadProtocol&, lldb_private::RegisterInfoInterface*)
2021-06-15T08:13:15.3977072Z    NativeRegisterContextRegisterInfo(
2021-06-15T08:13:15.3977759Z    ^
2021-06-15T08:13:15.3979398Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h:24:3: note:   candidate expects 2 arguments, 0 provided
2021-06-15T08:13:15.3982958Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:296:34: error: use of deleted function ‘lldb_private::NativeRegisterContextWatchpoint_x86::NativeRegisterContextWatchpoint_x86()’
2021-06-15T08:13:15.3985029Z        m_reg_info(), m_gpr_x86_64() {
2021-06-15T08:13:15.3985435Z                                   ^
2021-06-15T08:13:15.3986825Z In file included from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.h:15:0,
2021-06-15T08:13:15.3988829Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:11:
2021-06-15T08:13:15.3992372Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextWatchpoint_x86.h:16:7: note: ‘lldb_private::NativeRegisterContextWatchpoint_x86::NativeRegisterContextWatchpoint_x86()’ is implicitly deleted because the default definition would be ill-formed:
2021-06-15T08:13:15.3994803Z  class NativeRegisterContextWatchpoint_x86
2021-06-15T08:13:15.3995550Z        ^
2021-06-15T08:13:15.3998457Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextWatchpoint_x86.h:16:7: error: no matching function for call to ‘lldb_private::NativeRegisterContextRegisterInfo::NativeRegisterContextRegisterInfo()’
2021-06-15T08:13:15.4001483Z In file included from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Linux/NativeRegisterContextLinux.h:12:0,
2021-06-15T08:13:15.4003514Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.h:14,
2021-06-15T08:13:15.4005483Z                  from /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:11:
2021-06-15T08:13:15.4009298Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h:24:3: note: candidate: lldb_private::NativeRegisterContextRegisterInfo::NativeRegisterContextRegisterInfo(lldb_private::NativeThreadProtocol&, lldb_private::RegisterInfoInterface*)
2021-06-15T08:13:15.4012018Z    NativeRegisterContextRegisterInfo(
2021-06-15T08:13:15.4012712Z    ^
2021-06-15T08:13:15.4014336Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/./Plugins/Process/Utility/NativeRegisterContextRegisterInfo.h:24:3: note:   candidate expects 2 arguments, 0 provided
2021-06-15T08:13:15.4018478Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp: In member function ‘virtual llvm::Optional<lldb_private::process_linux::NativeRegisterContextLinux::SyscallData> lldb_private::process_linux::NativeRegisterContextLinux_x86_64::GetSyscallData()’:
2021-06-15T08:13:15.4022395Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1043:50: error: could not convert ‘(const uint8_t*)(& Int80)’ from ‘const uint8_t* {aka const unsigned char*}’ to ‘llvm::ArrayRef<unsigned char>’
2021-06-15T08:13:15.4023958Z      return SyscallData{Int80, Args, lldb_eax_i386};
2021-06-15T08:13:15.4024496Z                                                   ^
2021-06-15T08:13:15.4026769Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1043:50: error: could not convert ‘(const uint32_t*)(& Args)’ from ‘const uint32_t* {aka const unsigned int*}’ to ‘llvm::ArrayRef<unsigned int>’
2021-06-15T08:13:15.4030048Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1050:54: error: could not convert ‘(const uint8_t*)(& Syscall)’ from ‘const uint8_t* {aka const unsigned char*}’ to ‘llvm::ArrayRef<unsigned char>’
2021-06-15T08:13:15.4031630Z      return SyscallData{Syscall, Args, lldb_rax_x86_64};
2021-06-15T08:13:15.4032199Z                                                       ^
2021-06-15T08:13:15.4034643Z /tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1050:54: error: could not convert ‘(const uint32_t*)(& Args)’ from ‘const uint32_t* {aka const unsigned int*}’ to ‘llvm::ArrayRef<unsigned int>’
2021-06-15T08:13:15.4037663Z tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/build.make:176: recipe for target 'tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_x86_64.cpp.o' failed
2021-06-15T08:13:15.4040167Z make[2]: *** [tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_x86_64.cpp.o] Error 1
2021-06-15T08:13:15.4041894Z make[2]: Leaving directory '/tmp/llvm-20210615-5666-1pbq65d/llvm-project-12.0.0.src/llvm/build'
2021-06-15T08:13:15.4043675Z CMakeFiles/Makefile2:52365: recipe for target 'tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/all' failed
2021-06-15T08:13:15.4045177Z make[1]: *** [tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/all] Error 2
2021-06-15T08:13:15.4046126Z make[1]: *** Waiting for unfinished jobs....
```

**Comment by danielnachun** [comment]
So it turns out that this builds and passes all test just fine if I disable `lldb`.  This makes me think that a small patch or some other flag is all that is needed to make this build with gcc-5.  I'll keep looking into this while I upstream what we have so far to homebrew/core.  The changes we would need to make to build this with GCC 5 should be minor compared to what already needs to be upstreamed for everything else we've changed.

**Comment by danielnachun** [comment]
Found this commit that looks like it could fix it: https://github.com/llvm/llvm-project/commit/3bea7306e8669f94bacafae68748a9139cfc0b98#diff-74961116a1170ae453fe5fe92ebbbe08062fad015a1ffac609b9f61cd2b55014.  I'll test it as a patch locally and apply if it fixes the build.

**Comment by chandlerc** [comment]
Catching up here, nice!

> Found this commit that looks like it could fix it: [llvm/llvm-project@3bea730#diff-74961116a1170ae453fe5fe92ebbbe08062fad015a1ffac609b9f61cd2b55014](https://github.com/llvm/llvm-project/commit/3bea7306e8669f94bacafae68748a9139cfc0b98#diff-74961116a1170ae453fe5fe92ebbbe08062fad015a1ffac609b9f61cd2b55014). I'll test it as a patch locally and apply if it fixes the build.

Yeah, this should have been pulled into the branch. I've sent a note asking for that to happen so that if we get a point release, maybe eventually it has this fixed.

**Comment by iMichka** [comment]
There was a small issue with the workflow, fixed it here: https://github.com/Homebrew/linuxbrew-core/commit/a1ad4dd920d0b7d5bbc963f660fdb9b3858116fb

Can you rebase?

**Comment by danielnachun** [comment]
I rebased but I think the error is from the `libelf` postinstall failing.  I don't really understand the error I get when I run `brew postinstall -v --debug libelf`:

```
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall libelf
==> No available formula with the name "/home/linuxbrew/.linuxbrew/opt/libelf/.brew/libelf.rb".
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:364:in `get_formula'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formulary.rb:410:in `factory'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula_installer.rb:1065:in `post_install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/postinstall.rb:30:in `block in postinstall'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/postinstall.rb:27:in `each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/postinstall.rb:27:in `postinstall'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:122:in `<main>'
```

**Comment by danielnachun** [comment]
@chandlerc I was able to backport the commit I linked to the 12.0.0 release.  I couldn't use the commit directly because some file names have changed, but I modified it and attached it as a patch to the formula.  While that does appear to have fixed the failures I encountered before, we seem to have another unrelated issue, also while building `lldb`:
```
2021-06-18T18:11:48.2354322Z /tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp: In member function ‘virtual llvm::Optional<lldb_private::process_linux::NativeRegisterContextLinux::SyscallData> lldb_private::process_linux::NativeRegisterContextLinux_x86_64::GetSyscallData()’:
2021-06-18T18:11:48.2357782Z /tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1045:50: error: could not convert ‘(const uint8_t*)(& Int80)’ from ‘const uint8_t* {aka const unsigned char*}’ to ‘llvm::ArrayRef<unsigned char>’
2021-06-18T18:11:48.2359186Z      return SyscallData{Int80, Args, lldb_eax_i386};
2021-06-18T18:11:48.2359632Z                                                   ^
2021-06-18T18:11:48.2361617Z /tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1045:50: error: could not convert ‘(const uint32_t*)(& Args)’ from ‘const uint32_t* {aka const unsigned int*}’ to ‘llvm::ArrayRef<unsigned int>’
2021-06-18T18:11:48.2364747Z /tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1052:54: error: could not convert ‘(const uint8_t*)(& Syscall)’ from ‘const uint8_t* {aka const unsigned char*}’ to ‘llvm::ArrayRef<unsigned char>’
2021-06-18T18:11:48.2366114Z      return SyscallData{Syscall, Args, lldb_rax_x86_64};
2021-06-18T18:11:48.2366590Z                                                       ^
2021-06-18T18:11:48.2368552Z /tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp:1052:54: error: could not convert ‘(const uint32_t*)(& Args)’ from ‘const uint32_t* {aka const unsigned int*}’ to ‘llvm::ArrayRef<unsigned int>’
2021-06-18T18:11:48.2371195Z tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/build.make:176: recipe for target 'tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_x86_64.cpp.o' failed
2021-06-18T18:11:48.2373269Z make[2]: *** [tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_x86_64.cpp.o] Error 1
2021-06-18T18:11:48.2374740Z make[2]: Leaving directory '/tmp/llvm-20210618-5662-didxud/llvm-project-12.0.0.src/llvm/build'
2021-06-18T18:11:48.2376155Z CMakeFiles/Makefile2:52365: recipe for target 'tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/all' failed
2021-06-18T18:11:48.2377447Z make[1]: *** [tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/all] Error 2
2021-06-18T18:11:48.2378278Z make[1]: *** Waiting for unfinished jobs....
```
It seems perhaps that `gcc-5` doesn't understand how to handle `llvm::ArrayRef`?  If this is a bug that should be reported upstream, I'm happy to do so.

**Comment by chandlerc** [comment]
Yeah, the upstream fix for GCC-5 is incomplete. I've updated the LLVM bug with this: https://bugs.llvm.org/show_bug.cgi?id=50732

The added patch you need is:
```
diff --git a/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp b/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp
index db8dd00..e222909 100644
--- a/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp
+++ b/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_x86_64.cpp
@@ -1042,14 +1042,16 @@ NativeRegisterContextLinux_x86_64::GetSyscallData() {
     static const uint32_t Args[] = {lldb_eax_i386, lldb_ebx_i386, lldb_ecx_i386,
                                     lldb_edx_i386, lldb_esi_i386, lldb_edi_i386,
                                     lldb_ebp_i386};
-    return SyscallData{Int80, Args, lldb_eax_i386};
+    return SyscallData{llvm::makeArrayRef(Int80), llvm::makeArrayRef(Args),
+                       lldb_eax_i386};
   }
   case llvm::Triple::x86_64: {
     static const uint8_t Syscall[] = {0x0f, 0x05};
     static const uint32_t Args[] = {
         lldb_rax_x86_64, lldb_rdi_x86_64, lldb_rsi_x86_64, lldb_rdx_x86_64,
         lldb_r10_x86_64, lldb_r8_x86_64,  lldb_r9_x86_64};
-    return SyscallData{Syscall, Args, lldb_rax_x86_64};
+    return SyscallData{llvm::makeArrayRef(Syscall), llvm::makeArrayRef(Args),
+                       lldb_rax_x86_64};
   }
   default:
     llvm_unreachable("Unhandled architecture!");
```

The complete patch that builds with GCC 5 for me is also in the bug. Let me know if I can help more here.

**Comment by danielnachun** [comment]
I've tested the addition you made to the patch in our Docker environment and it works perfectly!  I am going to push the updated patch and this should be ready to go!  I'm also going to make a change to how to we do the stripping to use the `install/strip` target, which I've also tested and had no issues with, and clean up some syntax issues brought up in homebrew/core.

**Comment by iMichka** [comment]
Just a small heads up that there was stil a small issue in the bottle upload in our test workflow. This should be fixed now, but be aware that you will have to rebase this PR if you want to pull the bottle.

**Comment by danielnachun** [comment]
Thanks for the heads up, I rebased.

File: Formula/llvm.rb [hunk 1]
```diff
@@ -42,14 +32,13 @@ class Llvm < Formula
...
+    depends_on "glibc" if Formula["glibc"].any_version_installed?
     depends_on "pkg-config" => :build
     depends_on "binutils" # needed for gold
     depends_on "libelf" # openmp requires <gelf.h>
-  end
 
-  depends_on "gcc" unless OS.mac?
-
-  fails_with gcc: "5"
+    patch :DATA
```
Comment (chandlerc) [inline Formula/llvm.rb:40]: Maybe leave a comment so folks know where this patch came from and how to update it if needed?

```suggestion
    # Apply patches slated for the 12.0.x release stream
    # to allow building with GCC 5 and 6. Upstream bug:
    # https://bugs.llvm.org/show_bug.cgi?id=50732
    patch :DATA
```

**Review by chandlerc — APPROVED** [review]
LGTM for what its worth. A suggestion of a comment below, but feel free to do whatever makes sense for you all here. =D

File: Formula/llvm.rb (hunk 1 above)
Comment (danielnachun) [reply Formula/llvm.rb:40]: This comment will be very helpful when we upstream to homebrew/core and I'll add it once I get a green CI from the current run.

**Comment by danielnachun** [comment]
I'm going to publish and commit this bottle now that everything is green and add the comment in a separate PR since it will have no effect on the build process.

**Pull request closed without merging**